---
title: "Additional file 11 -  Plot of _ROH_ like Figure 1 for the advanced filter"
author:
date:
output:
  pdf_document: default
  html_document: default
---
\renewcommand{\figurename}{Figure}
\renewcommand{\thefigure}{S11.\arabic{figure}}
\renewcommand{\tablename}{Table}
\renewcommand{\thetable}{S11.\arabic{table}}

```{r setup, include=FALSE}

library(ggplot2)
library(readr)
library(dplyr)
library(patchwork)
library(knitr)

# Set default echo = FALSE for all code chunks - to only list the results, not the code
knitr::opts_chunk$set(echo = FALSE)

# Set the default working directory for all code chunks
knitr::opts_knit$set(root.dir = '/yourdata/autosomes')

# Set theme used for plotting
theme_hg <- theme_bw 
theme_set(theme_hg(base_size = 12))


```


```{r}

# Set filter text for a fixed filter - the advanced filter 
filter_text <- "FILTER='PASS' & QUAL>30 & FMT/DPU>10"

# Set file name of ROH report saved to disk from Addfile6 using the merged_vcf (pass.minQ30.minDPU10.snps.mergedsample.auto.vcf.gz) with the advanced filter

rohreport_name <- "rohreportAddfile6.csv"

```


```{r}

# Create a table for the filter-text
filter_table <- tibble(
  `Autosome files` = "SNP VCFs",
  `Filter-text` = filter_text
)


# Output the filter-text table
kable(filter_table,
       caption = "Filter (`bcftools filter -i`) used to filter the SNP files before further processing with PLINK." )

```


```{r}

# Read in the ROH data previously generated
rohreport <- read_csv(rohreport_name,
                           show_col_types = FALSE)

# Check output
# rohreport

```

```{r}

# Generate the plot of ROH KB/1000
p_roh <- 
  ggplot(rohreport, aes(x=Mb, y = KB/1000)) +
  # geom_hline(aes(yintercept = 0.007*2881.033286, linetype = "0.007", colour = "0.007")) +
  # geom_hline(aes(yintercept = 0.258*2881.033286, linetype = "0.258", colour = "0.258")) +
  geom_hline(aes(yintercept = 0.010*2881.033286, linetype = "0.010", colour = "0.010")) +
  geom_hline(aes(yintercept = 0.26*2881.033286, linetype = "0.260", colour = "0.260")) +
  geom_boxplot(aes(group = Mb), width = 0.05, size=0.1, outlier.size=0.1) + 
  scale_linetype_manual(name = expression(italic(F[ROH])), values = c(2, 2, 2)) +
  scale_colour_manual(name = expression(italic(F[ROH])), values = c("black", "gray50", "gray")) +
  scale_y_continuous(# Add a second axis and specify its features
    sec.axis = sec_axis(trans=~.*1/2881.033286,
                        name= expression(italic(F[ROH]))),
    breaks = seq(0, 800, 200)
  ) +
  scale_x_continuous(breaks = seq(0, 5, 0.5)) +
  xlab("Min length (Mb)") +
  ylab(expression(paste(italic(ROH), " ",italic(KB/1000)))) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())  +
  theme(legend.position = c(0.8, 0.55))

# Check output
# p_roh

```

```{r}

# Generate the plot of KBAVG/1000
p_rohavg <-  
  ggplot(rohreport, aes(x=Mb, y = KBAVG/1000)) +
  geom_boxplot(aes(group = Mb), width = 0.05, size=0.1, outlier.size=0.1) + 
  scale_y_continuous(breaks=seq(0, 14, 2)) +
  xlab("Min length (Mb)") +
  scale_x_continuous(breaks = seq(0, 5, 0.5)) +
  ylab(expression(italic(KBAVG/1000))) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) 

# Check output
# p_rohavg

```


```{r, fig.width = 7, fig.asp = 0.85, out.width="80%", fig.align="center", fig.cap = "$ROH$ of the eight merged genomes (autosomes) from the Faroe Islands. $a,b$ Total $ROH=KB/1000$ and average $ROH=KBAVG/1000$ are the total and average length of $ROH$, respectively, in the PLINK --homozyg report for variable minimum segment length --homozyg-kb. This we varied from $100$ to $5000$kb in $100$kb increments. $a$ Boxplots of the total $ROH$ (left axis) and of the inbreeding coefficient $F_{ROH} = KB/L$ (right axis) in which  $L = 2881033.286$kb is the autosome length. The median inbreeding $F_{ROH>5}=0.010$ (black dashed line) is the lowest level of recent inbreeding, which is like the average pedigree inbreeding of $0.0067$ and $0.0081$ estimated for multiple scelerosis patients and controls, respectively, from the Faroese population. The $F_{ROH>0.1}=0.260$  (gray dashed line) is the highest level of both recent and ancient inbreeding, which is like $F_{ROH>0.1}$ for the European population of the 1000GP."}

# Combine and label the two plots by using the patchwork package and output the plot
((p_roh + labs(tag = 'a')) / (p_rohavg + labs(tag = 'b'))) & theme(plot.tag = element_text(face = 'bold'))
```







